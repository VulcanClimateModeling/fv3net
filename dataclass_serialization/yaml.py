from typing import Type
import yaml
import dacite
import dataclasses
import dacite.types


def _make_repr(
    dumper: Type[yaml.SafeDumper],
    cls,
    tag=yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
):
    names = [field.name for field in dataclasses.fields(cls)]

    def _represent(dumper: yaml.SafeDumper, data: cls):
        mapping = {}
        for name in names:
            mapping[name] = getattr(data, name)

        return dumper.represent_mapping(tag, mapping)

    dumper.add_representer(cls, _represent)


class Visitor:
    def visit_dataclass(self, cls):
        if cls in self.visited:
            return

        _make_repr(self.dumper, cls)
        self.visited.add(cls)

    def visit_leaf(self, cls):
        pass

    def visit_union(self, cls):
        pass


class DumperVisitor(Visitor):
    def __init__(self, dumper):
        self.visited = set()
        self.dumper = dumper

    def visit_dataclass(self, node: "DataclassNode"):
        if node.cls in self.visited:
            return

        tag = (
            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG
            if node.name
            else "!" + node.cls.__name__
        )

        _make_repr(
            self.dumper,
            node.cls,
            tag,
        )
        self.visited.add(node.cls)


class LoaderVisitor(Visitor):
    def __init__(self, loader: Type[yaml.SafeLoader]):
        self.visited = set()
        self.loader = loader

    def visit_dataclass(self, node):
        cls = node.cls
        if cls in self.visited:
            return

        def _constructor(loader: yaml.SafeLoader, node):
            mapping = loader.construct_mapping(node)
            return cls(**mapping)

        if not node.name:
            tag = "!" + cls.__name__
            self.loader.add_constructor(tag, _constructor)
            self.visited.add(cls)


def to_yaml(obj):
    dumper = yaml.SafeDumper
    cls = obj.__class__
    type_graph = build_type_tree("top", cls)
    visitor = DumperVisitor(dumper)
    for node in walk_post_order(type_graph):
        node.accept(visitor)
    return yaml.dump(obj, Dumper=dumper)


def from_yaml(cls, yaml_str):
    loader = yaml.SafeLoader
    visitor = LoaderVisitor(loader)
    type_graph = build_type_tree("top", cls)
    for node in walk_post_order(type_graph):
        node.accept(visitor)

    return dacite.from_dict(cls, yaml.load(yaml_str, Loader=loader))


def is_union(cls: Type) -> bool:
    return dacite.types.is_union(cls)


class Node:
    def __init__(self, name, cls, children):
        self.name = name
        self.cls = cls
        self.children = children


class UnionNode(Node):
    def accept(self, visitor):
        visitor.visit_union(self)


class GenericNode(Node):
    def accept(self, visitor):
        visitor.visit_generic(self)


class DataclassNode(Node):
    def accept(self, visitor):
        visitor.visit_dataclass(self)


class OtherNode(Node):
    def accept(self, visitor):
        visitor.visit_leaf(self)


def walk_post_order(node: Node):
    stack = [(True, node)]
    while stack:
        first_visit, node = stack.pop()

        if first_visit:
            stack.append((False, node))
            for kids in node.children:
                stack.append((first_visit, kids))
        else:
            yield node


def build_type_tree(name, cls):
    if dataclasses.is_dataclass(cls):
        return DataclassNode(
            name,
            cls,
            children=[
                build_type_tree(field.name, field.type)
                for field in dataclasses.fields(cls)
            ],
        )
    elif is_union(cls):
        return UnionNode(
            name,
            cls,
            [
                build_type_tree("", child)
                for k, child in enumerate(dacite.types.extract_generic(cls))
            ],
        )
    elif dacite.types.is_generic(cls):
        children = list(dacite.types.extract_generic(cls))
        return GenericNode(
            name, cls, [build_type_tree("", child) for child in children]
        )
    else:
        return OtherNode(name, cls, [])
