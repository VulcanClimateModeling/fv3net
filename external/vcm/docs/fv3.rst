Working with FV3 outputs
========================

FV3GFS restart files
--------------------

I/O
~~~

Restart files can be loaded from disk with the functions

.. autofunction:: vcm.fv3_restarts.open_restarts


Coarse-graining
~~~~~~~~~~~~~~~

VCM Tools provides two high-level methods for coarse-graining a complete set of
restart files produced from running FV3GFS.  The two methods perform the
coarse-graining in different ways.

.. autofunction:: vcm.cubedsphere.coarsen_restarts_on_sigma
.. autofunction:: vcm.cubedsphere.coarsen_restarts_on_pressure

Parsing FV3 logs
----------------

Use the following routines to parse statistical information from the console outputs of an FV3 run:

.. autofunction:: vcm.fv3.logs.loads


Standardizing FV3 outputs
-------------------------

The diagnostics output by FV3GFS have verbose, and sometimes difficult to
understand dimension names.  The functions here provide convenient ways for
renaming dimensions to and from standard names, as well as standardizing time
coordinate information:

.. automodule:: vcm.fv3.metadata
   :members:
   :undoc-members:
   :show-inheritance:
