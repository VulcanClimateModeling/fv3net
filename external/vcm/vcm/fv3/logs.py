#!/usr/bin/env python
import re
from typing import Mapping, List, Tuple
from collections import defaultdict
from datetime import datetime
from toolz import curry
import dataclasses


__all__ = ["loads"]


def parse_date_stats_block(line):
    tokens = line.split()
    args = [int(tok) for tok in tokens]
    return datetime(*args)


@curry
def parse_pattern(pattern, line):
    water_vapor_regex = re.compile(pattern + r" *= (.*)")
    match = water_vapor_regex.search(line)
    if match:
        return float(match.group(1).strip())
    else:
        return None


def set_initial_date(dates):
    if dates[0] is None:
        initial_date = dates[1] - (dates[2] - dates[1])
        return [initial_date] + dates[1:]
    else:
        return dates


@dataclasses.dataclass
class FV3Log:
    dates: List[datetime]
    totals: Mapping[str, List[float]]
    ranges: Mapping[str, List[Tuple[float, float]]]


def loads(log: str) -> FV3Log:
    """Parse the statistics information from the logs for an FV3 simulation

    Does not parse python-related diagnostics. It extracts the statistical
    information from blocks with the following format::

                2016           8           1           1           0           0
        ZS      6849.180      -412.0000       231.8707
        PS max =    1052.038      min =    439.9279
        Mean specific humidity (mg/kg) above 75 mb=   3.993258
        Total surface pressure (mb) =    985.9462
        mean dry surface pressure =    983.2382
        Total Water Vapor (kg/m**2) =   27.51812
        --- Micro Phys water substances (kg/m**2) --
        Total cloud water=  3.4418099E-0
        Total rain  water=  8.9859152E-0
        Total cloud ice  =  3.6711354E-0
        Total snow       =  1.4170707E-0
        Total graupel    =  1.3332913E-0
        --------------------------------------------
        TE ( Joule/m^2 * E9) =   2.633417
        UA_top max =    138.3022      min =   -45.65562
        UA max =    138.3022      min =   -52.00678
        VA max =    60.31665      min =   -67.41895
        W  max =    36.12219      min =   -11.35321
        Bottom w max =    5.591572      min =   -6.320749
        Bottom: w/dz max =   0.2113777      min =  -0.2430508
        DZ (m) max =   -21.97151      min =   -5737.403
        Bottom DZ (m) max =   -21.97151      min =   -32.53793
        TA max =    319.3932      min =    171.7615
        OM max =    79.11412      min =   -91.14186
        ZTOP      40.79998       34.23963       39.11939
        SLP max =    1064.040      min =    953.2686
        ATL SLP max =    1022.554      min =    1010.739
        fv_GFS Z500   5700.141       5790.903       5459.689       5864.878
        Cloud_top_P (mb) max =   1.0000000E+10  min =    14.69188
        Surf_wind_speed max =    33.00807      min =   4.1885415E-0
        T_bot:      319.3932       216.0579       289.6649
        ZTOP max =    40.79998      min =    34.23963
        W5km max =    13.85566      min =   -5.952400
        max reflectivity =    64.90421      dBZ
        sphum max =   2.3002494E-02  min =   9.9999342E-10
        liq_wat max =   2.2458597E-03  min =  -1.3747283E-19
        rainwat max =   8.2634874E-03  min =  -1.4335402E-18
        ice_wat max =   5.2885562E-03  min =  -5.2392893E-20
        snowwat max =   8.4350556E-03  min =  -9.2761843E-19
        graupel max =   1.3989445E-02  min =  -4.0843832E-05
        o3mr max =   1.7852419E-05  min =   2.2205915E-09
        sgs_tke max =    130.6761      min =   9.8705533E-10

    Currently it only supports the min/max and total statistics.

    Returns:
        fv3log: an object with the parsed statistical information
        
    
    """
    lines = log.splitlines()
    min_max_regex = re.compile(r"(.*)max *= *([0-9E\-\.]+) *min = *([0-9E\-\+\.]+)")
    zs_regexp = re.compile(r"ZS *([0-9E\-\.]+) +([0-9E\-\.]+) +([0-9E\-\+\.]+)")
    totals = defaultdict(list)
    min_max = defaultdict(list)
    dates = []

    mean_patterns = {
        "total surface pressure": parse_pattern(r"Total surface pressure \(mb\)"),
        "total water vapor": parse_pattern(r"Total Water Vapor \(kg/m\*\*2\)"),
        "mean dry surface pressure": parse_pattern("mean dry surface pressure"),
        "total cloud water": parse_pattern("Total cloud water"),
        "total rain water": parse_pattern(r"Total rain\s+water"),
        "total cloud ice": parse_pattern("Total cloud ice"),
        "total snow": parse_pattern("Total snow"),
        "total graupel": parse_pattern("Total graupel"),
    }

    date = None
    for line in lines:

        if "fv_restart_end" in line:
            # the logs end with a block like this::
            #
            # fv_restart_end u    =    5404162460055465614
            # ...
            #  ZS   6849.180      -412.0000       231.8706
            #
            # This break avoids appending the final date twice.
            break

        try:
            date = parse_date_stats_block(line)
        except (TypeError, ValueError):
            pass

        match = zs_regexp.search(line)
        if match:
            # The statistics block begins with the date followed by ZS::
            #         2016           8           2           8           0           0
            #  ZS      6849.180      -412.0000       231.8707
            dates.append(date)

        match = min_max_regex.search(line)
        if match:
            name, max_, min_ = match.groups()
            name = name.strip()
            min_max[name].append((float(min_), float(max_)))

        for variable, parser in mean_patterns.items():
            val = parser(line)
            if val is not None:
                totals[variable].append(val)

    return FV3Log(set_initial_date(dates), dict(totals), dict(min_max))
