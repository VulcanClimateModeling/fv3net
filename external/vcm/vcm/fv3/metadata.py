import xarray as xr
from typing import Mapping, Set, Callable, Sequence
import logging
from ..convenience import round_time
from ..safe import warn_if_intersecting

logger = logging.getLogger(__name__)


DIM_RENAME_INVERSE_MAP = {
    "x": {"grid_xt", "grid_xt_coarse"},
    "y": {"grid_yt", "grid_yt_coarse"},
    "tile": {"rank"},
    "x_interface": {"grid_x", "grid_x_coarse"},
    "y_interface": {"grid_y", "grid_y_coarse"},
}
VARNAME_SUFFIX_TO_REMOVE = ["_coarse"]
TIME_DIM_NAME = "time"
STANDARD_TO_GFDL_DIM_MAP = {
    "x_interface": "grid_x",
    "y_interface": "grid_y",
    "x": "grid_xt",
    "y": "grid_yt",
    "z": "pfull",
}


def standardize_fv3_diagnostics(
    ds: xr.Dataset, time: str = TIME_DIM_NAME
) -> xr.Dataset:
    """Standardize dimensions, coordinates, and attributes of FV3 diagnostic output
    
    Args:
        ds (xr.Dataset):
            Dataset of FV3GFS or SHiELD diagnostics outputs (either Fortran or
            python-derived), presumably opened from disk via zarr.
        time (str, optional):
            Name of the time coordinate dimension in the dataset
        
    Returns:
        The dataset with coordinate names and tile ranges set to CF "standard" values,
        variable name suffixes (e.g., "_coarse") removed, and data variable attributes
        ("long_name" and "units") set. The time coordinate is rounded to the nearest
        second.
    
    """

    if time in ds.coords:
        ds[time].attrs["calendar"] = "julian"

    funcs: Sequence[Callable[[xr.Dataset], xr.Dataset]] = [
        xr.decode_cf,
        _adjust_tile_range,
        _rename_dims,
        _round_time_coord,
        _remove_name_suffix,
        _set_missing_attrs,
    ]

    for func in funcs:
        ds = func(ds)

    return ds


def _adjust_tile_range(ds: xr.Dataset) -> xr.Dataset:

    if "tile" in ds:
        tiles = ds.tile

        if tiles.isel(tile=-1) == 6:
            ds = ds.assign_coords({"tile": tiles - 1})

    return ds


def _rename_dims(
    ds: xr.Dataset, rename_inverse: Mapping[str, Set[str]] = DIM_RENAME_INVERSE_MAP
) -> xr.Dataset:

    varname_target_registry = {}
    for target_name, source_names in rename_inverse.items():
        varname_target_registry.update({name: target_name for name in source_names})

    vars_to_rename = {
        var: varname_target_registry[str(var)]
        for var in ds.dims
        if var in varname_target_registry
    }
    ds = ds.rename(vars_to_rename)
    return ds


def _round_time_coord(ds: xr.Dataset, time_coord: str = TIME_DIM_NAME) -> xr.Dataset:

    if time_coord in ds.coords:
        new_times = round_time(ds[time_coord])
        ds = ds.assign_coords({time_coord: new_times})
    else:
        logger.debug(
            "Round time operation called on dataset missing a time coordinate."
        )

    return ds


def _set_missing_attrs(ds: xr.Dataset) -> xr.Dataset:

    for var in ds:
        da = ds[var]

        # True for some prognostic zarrs
        if "description" in da.attrs and "long_name" not in da.attrs:
            da.attrs["long_name"] = da.attrs["description"]

        if "long_name" not in da.attrs:
            da.attrs["long_name"] = var

        if "units" not in da.attrs:
            da.attrs["units"] = "unspecified"
    return ds


def _remove_name_suffix(
    ds: xr.Dataset, suffixes: Sequence[str] = VARNAME_SUFFIX_TO_REMOVE
) -> xr.Dataset:
    for target in suffixes:
        replace_names = {
            vname: str(vname).replace(target, "")
            for vname in ds.data_vars
            if target in str(vname)
        }

        warn_if_intersecting(ds.data_vars.keys(), replace_names.values())
        ds = ds.rename(replace_names)
    return ds


def gfdl_to_standard(ds: xr.Dataset):
    """Convert from GFDL dimension names (grid_xt, etc) to "standard"
    names (x, y, z)
    """

    key, val = STANDARD_TO_GFDL_DIM_MAP.keys(), STANDARD_TO_GFDL_DIM_MAP.values()
    inverse = dict(zip(val, key))

    return ds.rename({key: val for key, val in inverse.items() if key in ds.dims})


def standard_to_gfdl(ds: xr.Dataset):
    """Convert from "standard" names to GFDL names"""
    return ds.rename(
        {key: val for key, val in STANDARD_TO_GFDL_DIM_MAP.items() if key in ds.dims}
    )
