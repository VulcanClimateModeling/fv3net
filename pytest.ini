[pytest]
norecursedirs =
    data
    build
    external/fv3config
    external/radiation/tests/test_driver
    external/artifacts
    external/fv3fit
    external/fv3gfs-fortran
    external/fv3kube
    external/vcm
    external/emulation
    workflows/dataflow/tests/integration
    docker
    workflows/prognostic_c48_run
    workflows/post_process_run

markers =
    regression: marks regression tests (deselect with '-m "not regression"')
    network: marks tests requiring network access
    slow: marks tests as slow (approx >1s) (deselect with '-m "not slow"')
    scream_run: marks tests that require running inside scream container (delesect with '-m "not scream_run"')

filterwarnings =
    ignore:distutils Version classes are deprecated:DeprecationWarning
    ignore:Call to deprecated create function:DeprecationWarning
    ignore:.*is deprecated and will be removed in Pillow 10:DeprecationWarning
