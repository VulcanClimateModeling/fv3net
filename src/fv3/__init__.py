from .docker import make_experiment, run_experiment
from .coarsen import coarsen_sfc_data_in_directory
