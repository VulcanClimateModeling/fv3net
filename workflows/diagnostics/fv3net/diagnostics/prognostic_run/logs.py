from typing import List
import re
import json
import datetime


def parse_duration(logs: List[str]) -> datetime.timedelta:
    log = "".join(logs)
    profile_info = re.findall(r"INFO:profiles:(.*)", log)
    profile_json = [json.loads(s) for s in profile_info]
    profile_time = sorted(
        set([datetime.datetime.fromisoformat(data["time"]) for data in profile_json])
    )
    last_time = profile_time[-1]
    initial_time = profile_time[0] - (profile_time[1] - profile_time[0])
    duration = last_time - initial_time
    return duration
