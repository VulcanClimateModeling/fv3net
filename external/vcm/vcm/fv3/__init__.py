from . import logs
from .metadata import standardize_fv3_diagnostics, gfdl_to_standard, standard_to_gfdl
