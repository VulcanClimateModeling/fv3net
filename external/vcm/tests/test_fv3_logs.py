from vcm.fv3.logs import loads, parse_date_forecast_date
from datetime import datetime
from pathlib import Path
import pytest
import yaml


@pytest.fixture(params=["shield", "fv3net"], scope="session")
def shield_log(request):
    if request.param == "shield":
        path = (
            Path(__file__).parent
            / "C3072_0801_n11a_vulcan-2020-05-21-production-simulation.o268642462"
        )
        with open(path) as f:
            str_ = f.read()
        return str_
    else:
        path = Path(__file__).parent / "prognostic-run-logs.txt"
        with open(path) as f:
            str_ = f.read()
        return str_


def test_parse_date():
    example_date = "0 FORECAST DATE          14 AUG.  2016 AT 18 HRS  0.00 MINS"
    assert parse_date_forecast_date(example_date) == datetime(
        year=2016, month=8, day=14, hour=18
    )


def test_loads_with_shield_data(shield_log: str):
    fv3log = loads(shield_log)

    for key in fv3log.totals:
        assert len(fv3log.totals[key]) == len(fv3log.dates)


def test_loads_with_shield_data_date(shield_log: str, regtest):
    fv3log = loads(shield_log)
    yaml.safe_dump(fv3log.dates, regtest)


@pytest.mark.parametrize(
    "variable_name",
    [
        "total surface pressure",
        "total water vapor",
        "mean dry surface pressure",
        "total cloud water",
        "total rain water",
        "total cloud ice",
        "total snow",
        "total graupel",
    ],
)
def test_loads_shield_totals(shield_log: str, regtest, variable_name: str):
    fv3log = loads(shield_log)
    yaml.safe_dump(fv3log.totals[variable_name], regtest)
