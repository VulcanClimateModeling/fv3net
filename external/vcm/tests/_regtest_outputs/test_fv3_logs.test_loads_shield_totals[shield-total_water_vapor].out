- 27.45369
- 27.51812
- 27.564
- 27.59198
- 27.60608
- 27.60825
- 27.60315
- 27.59381
- 27.57987
- 27.56137
- 27.53864
